### PR TITLE
Add configuration to replace sections of path strings according to regexes, and tags spans accordingly

### PR DESCRIPTION
**Pull request by aideencostello**
Add configuration to replace sections of path strings according to regexes, and tags spans accordingly

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,14 +83,27 @@ We recommend using this feature in a server that has `OpenTracingServer` install
...
     filter { call -> call.request.path().startsWith("/_probes") }
 }
 ```
 
+### tagAndReplace
+When an API path contains some kind of id, it can be helpful to replace it with a constant string, so that similar spans
+are named the same and grouped together. The contents of the id is then tagged so that information is not lost.
+```kotlin
+install(OpenTracingServer) {
+    tagAndReplace("customId", Regex("""[0-9]{8}-[0-9]{4}"""))
```
Comment (hj1611) [inline README.md:100]: Not sure the name describes exactly what the function is doing here. What would you think of `extractTagFromPath` or `groupPathBy`?

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        val tagsToMatch = mapOf(Pair("customId", """[0-9]{8}-[0-9]{4}""".toRegex()))
+
+        val pathAndTags = path.toPathAndTags(tagsToMatch)
+
+        assertThat(pathAndTags.path).isEqualTo("/path/<customId>/")
+        assertThat(pathAndTags.tags).isEqualTo(mapOf(Pair("customId", "12345678-1234")))
+    }
+
+    @Test
+    fun `toPathAndTags can tag and replace multiple strings`() {
```
Comment (hj1611) [inline src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt:99]: What happens if you match a UUID with a customId?
If you start from
```kotlin
/path/dd988515-936b-4ecd-a6fb-03d74b424f49/hello
```
with
```kotlin
val tagsToMatch = mapOf(
    Pair("customId", """<UUID regex string>""".toRegex()),
)
```
will you get
```kotlin
/path/<customId>/hello`
```

or
```kotlin
/path/<UUID>/hello
```
?

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt [hunk 3]
```diff
@@ -47,22 +53,24 @@ class OpenTracingServer(
...
 
                 val headers: MutableMap<String, String> = call.request.headers.toMap()
                 headers.remove("Authorization")
 
                 val clientSpanContext: SpanContext? = tracer.extract(Format.Builtin.HTTP_HEADERS, TextMapAdapter(headers))
                 if (clientSpanContext == null) log.info("Tracing context could not be found in request headers. Starting a new server trace.")
 
-                val pathUuid: PathUuid = context.request.path().UuidFromPath()
-                val spanName = "${context.request.httpMethod.value} ${pathUuid.path}"
+                val pathAndTags = context.request.path().toPathAndTags(feature.toBeTaggedAndReplaced)
```
Comment (hj1611) [inline src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:64]: Here you can make the code a little more readable by destructuring the response:
```kotlin
val (path, tags) = context.request.path().toPathAndTags(feature.toBeTaggedAndReplaced)
```

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt (hunk 2 above)
Comment (aideencostello) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt:99]: I think I'd consider that a config error - you could get either with the current implementation as there's no order 🤷 

I don't really see how you can prevent people from having potentially overlapping regexes.

Are you saying you'd prefer to have an order so that we can say what the expected behaviour would be?

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt (hunk 3 above)
Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:64]: Oooh forgot you could do this in Kotlin, that's a lot nicer 👍

**Comment by aideencostello** [comment]
@hj1611 Somewhat related to your comment, you could say we are missing a feature of being able to anchor your regex.

The scenario I'm thinking of is if you have the same customId in multiple paths, and you want to call it different things according to the path:

`/path/butterfly/1234-4567/type` -> `/path/butterfly/<butterfly-id>/type`

`/path/cow/1245-4322/age` -> `/path/cow/<cow-id>/age`

I suspect this is overkill, but shout anyone if you think it would be useful.

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt [hunk 4]
```diff
@@ -21,22 +21,28 @@ import java.util.Stack
...
     class Configuration {
         val filters = mutableListOf<(ApplicationCall) -> Boolean>()
+        val tagsToExtractFromPath: MutableMap<String, Regex> = mutableMapOf(uuidTagAndReplace)
 
         fun filter(predicate: (ApplicationCall) -> Boolean) {
             filters.add(predicate)
         }
+
+        fun extractTagFromPath(tagName: String, regex: Regex) {
+            tagsToExtractFromPath[tagName] = regex
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:36]: minor; can we have the regex as the first parameter? Seems more idiomatic , as in String.replace

Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:36]: Was trying to be consistent throughout the code `extractTagFromPath(tagName, regex)` -> `Pair(tagName, regex)` -> `Pair(tagName, tagValue)` -> `spanBuilder.withTag(tagName, tagValue)`

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt [hunk 5]
```diff
@@ -10,14 +10,23 @@ import io.opentracing.propagation.Format
...
+class OpenTracingClient(
+        val tagsToExtractFromPath: Map<String, Regex>
+) {
+    class Configuration {
+        val tagsToExtractFromPath: MutableMap<String, Regex> = mutableMapOf(uuidTagAndReplace)
+
+        fun extractTagFromPath(tagName: String, regex: Regex) {
+            tagsToExtractFromPath[tagName] = regex
+        }
+    }
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt:22]: I would prefer if we don't add config to the client feature. It's a better experience for users to have all config in one place. You can have an internal variable (similar to 'threadLocalSpanStack`) that you can write to in the server feature and read from here in the client feature.

Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt:22]: What if your application isn't a ktor server?

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt [hunk 6]
```diff
@@ -21,22 +21,28 @@ import java.util.Stack
...
 ) {
     class Configuration {
         val filters = mutableListOf<(ApplicationCall) -> Boolean>()
+        val tagsToExtractFromPath: MutableMap<String, Regex> = mutableMapOf(uuidTagAndReplace)
 
         fun filter(predicate: (ApplicationCall) -> Boolean) {
             filters.add(predicate)
         }
+
+        fun extractTagFromPath(tagName: String, regex: Regex) {
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:35]: I prefer 'replace' over extract, which is less common. How about 'replaceInPathAndTagSpan' to be really explicit (this is complex config).

Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:35]: I started with replace and @hj1611 didn't like it 😂

File: CODEOWNERS [hunk 7]
```diff
@@ -1 +1 @@
-*   @fstien @hj1611
+*   @fstien @hj1611 @aideencostello
```
Comment (hj1611) [inline CODEOWNERS:1]: On the CODEOWNERS bit, I think it's best if we leave only @fstien as code owner, as he has the overview of the project, and whether something could break existing clients, he should get final approval of whether something is merged or not. Feel free to remove me from the list as well :)

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt (hunk 5 above)
Comment (fstien) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt:22]: Good point! The ScopeManager implementation requires the use of server feature unfortunately.

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt (hunk 4 above)
Comment (fstien) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:36]: I get you but the implementation shouldn't be influenced by what's on the interface here.

**Comment by fstien** [comment]
> @hj1611 Somewhat related to your comment, you could say we are missing a feature of being able to anchor your regex.
> 
> The scenario I'm thinking of is if you have the same customId in multiple paths, and you want to call it different things according to the path:
> 
> `/butterfly/1234-4567/type` -> `/butterfly/<butterfly-id>/type`
> 
> `/cow/1245-4322/age` -> `/cow/<cow-id>/age`
> 
> I suspect this is overkill, but shout anyone if you think it would be useful.
> 
> Edit: this could always be added later as an optional param

@hj1611  i see what you mean, it seems unlikely that you'd want a different string for the same regex. Users can always use a generic id string and should know their apps well enough to know what it's referring to (usually the id of the resource in the path haha)

**Comment by fstien** [comment]
@aideencostello Can you please give this PR a more descriptive name, which can be included in the release message as an announcement of the new feature.

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
+package com.zopa.ktor.opentracing
+
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import kotlinx.coroutines.runBlocking
+import org.junit.jupiter.api.Test
+
+class UtilsTest {
```
Comment (fstien) [inline src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt:8]: Can we please keep the test classes as they are and add the tests for this in KtorOpenTracingTest . I agree they need to be split but would rather do in a separate PR.

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt (hunk 4 above)
Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingServer.kt:36]: Yeah fair enough. I guess I was thinking of it as primarily a tag to be added, and you were thinking of it as primarily a replacement in the path 😁 I will switch it around

File: CODEOWNERS (hunk 7 above)
Comment (aideencostello) [reply CODEOWNERS:1]: I think that would make it impossible for @fstien to merge PRs

Comment (aideencostello) [reply CODEOWNERS:1]: (I was hoping to be able to get involved in reviewing @fstien's PRs)

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt (hunk 8 above)
Comment (aideencostello) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt:8]: Sure 👍 I'll pull it out into a separate PR

Comment (fstien) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt:8]: excellent , thanks ! :D

File: CODEOWNERS (hunk 7 above)
Comment (hj1611) [reply CODEOWNERS:1]: Oh don't worry you can still review prs if you're not a code owner, and merging PRs should work fine through admin access :)

File: README.md [hunk 9]
```diff
@@ -83,14 +83,27 @@ We recommend using this feature in a server that has `OpenTracingServer` install
...
+Your application might be serving static content (such as k8s probes), for which you do not want to create traces. 
 You can filter these out as follows:
 ```kotlin
 install(OpenTracingServer) {
     filter { call -> call.request.path().startsWith("/_probes") }
 }
 ```
 
+### replaceInPathAndTagSpan
+When an API path contains some kind of id, it can be helpful to replace it with a constant string, so that similar spans
```
Comment (fstien) [inline README.md:96]:
```suggestion
When a request path contains an id, you can replace it with a constant string in the span operation name. This ensures that requests for different ids have the same span operation name. 
```

File: README.md [hunk 10]
```diff
@@ -83,14 +83,27 @@ We recommend using this feature in a server that has `OpenTracingServer` install
...
+are named the same and grouped together. The contents of the id is then tagged so that information is not lost.
```
Comment (fstien) [inline README.md:97]: The value of the id is then tagged on the span.

File: README.md [hunk 11]
```diff
@@ -83,14 +83,27 @@ We recommend using this feature in a server that has `OpenTracingServer` install
...
+```kotlin
+install(OpenTracingServer) {
+    replaceInPathAndTagSpan(Regex("""[0-9]{8}-[0-9]{4}"""), "customId")
+}
+```
+In the above example, `/path/12345678-1234` would be recorded as `/path/<customId>` with the tag `customId=12345678-1234`.
```
Comment (fstien) [inline README.md:103]:
```suggestion
In the above example, `/path/12345678-1234` would lead to a span named as `/path/<customId>` with the tag `customId=12345678-1234`.
```

File: README.md [hunk 12]
```diff
@@ -83,14 +83,27 @@ We recommend using this feature in a server that has `OpenTracingServer` install
...
+
+Note that UUIDs are already tagged and replaced by default.
```
Comment (fstien) [inline README.md:105]:
```suggestion
Note that UUIDs are already tagged and replaced by default with `UUID`.
```

File: src/test/kotlin/com/zopa/ktor/opentracing/KtorOpenTracingTest.kt [hunk 13]
```diff
@@ -289,7 +290,7 @@ class KtorOpenTracingTest {
     }
 
     @Test
-    fun `Client passes trace context in headers`() {
+    fun `Client passes trace context in headers and tags uuids`() {
```
Comment (fstien) [inline src/test/kotlin/com/zopa/ktor/opentracing/KtorOpenTracingTest.kt:293]: this test is already very complex (not easy to assert on the client headers ) so can we maybe have this in a different test please

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt (hunk 2 above)
Comment (hj1611) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt]: I think I would personally prefer having the customId take priority here because it would be a frustrating experience for the user having the default overriding what they specified. Maybe throwing an exception when there is a clash could be an easy way to deal with this?

Comment (aideencostello) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt]: That makes sense. Are you happy with the following?

Config:
```kotlin
install(OpenTracingServer) {
    replaceInPathAndTagSpan(Regex(""), "A")
    replaceInPathAndTagSpan(Regex(""), "B")
}
```

Leading to a priority of:
* A
* B
* UUID

I don't think it's practical to throw an exception for clashes because you can so easily have regexes differ slightly but still perform the same function.

File: src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt (hunk 5 above)
Comment (aideencostello) [reply src/main/kotlin/com/zopa/ktor/opentracing/OpenTracingClient.kt]: Pushed dfc0640

File: src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt (hunk 2 above)
Comment (aideencostello) [reply src/test/kotlin/com/zopa/ktor/opentracing/UtilsTest.kt]: Pushed b81e87a

File: README.md [hunk 14]
```diff
@@ -85,14 +85,26 @@ We recommend using this feature in a server that has `OpenTracingServer` install
 
 ### Filter Requests
 
-Your application might be serving static content (such as k8s probes), for which you do not to create traces. 
+Your application might be serving static content (such as k8s probes), for which you do not want to create traces.
```
Comment (fstien) [inline README.md:88]: good catch
